serverlessworkflow/modules/ROOT/pages/use-cases/orchestration-based-saga-pattern.adoc

```diff
@@ -1 +1,269 @@
+               "failService":".failService"
+            }
+         },
+         "actionDataFilter":{
+            "results":".",
+            "toStateData":".paymentResponse"
+         }
+      }
+   ],
+   "transition":"scheduleShipping",
+   "compensatedBy":"CancelPayment",
+   "onErrors":[
+      {
+         "errorRef":"process payment failed",
+         "transition":"ServiceError"
+      }
+   ]
+}
+----
+
+=== Compensations
+
+The core functionality when designing the Saga is the compensation for each step to be executed by a participant.
+In serverless workflow, each state should define how it should be compensated using its compensatedBy property, which indicates another state responsible to perform the compensation for it, i.e. processPayment defines the CancelPayment as the compensation, as define in the <<payment-state, payment state>>.
+
+[source,json]
+"compensatedBy": "CancelPayment"
+
+=== Errors
+
+Compensations are triggered as soon as an error occurs, according to the errors definitions associated states on serverless workflow, where they are identified by a name, for instance, on `processPayment` there is the error `process payment failed` which is a reference to the <<#errors-definitions, errors section>> in the workflow definition.
+
+[#errors-definitions]
+.Example the error declaration for the Saga
+[source,json]
+----
+{
+   "errors":[
+      {
+         "name":"reserve stock failed",
+         "code":"org.kie.kogito.ServiceException"
+      },
+      {
+         "name":"process payment failed",
+         "code":"org.kie.kogito.ServiceException"
+      },
+      {
+         "name":"shipping failed",
+         "code":"org.kie.kogito.ServiceException"
+      }
+   ]
+}
+----
+
+The defined error uses the FQDN of Java exceptions thrown by functions, in the given example, `org.kie.kogito.ServiceException` is thrown by they service calls defined as a  <<#custom-function, Java method>> here is the link:{kogito_sw_examples_url}/serverless-workflow-saga-quarkus/src/main/java/org/kie/kogito/PaymentService.java[source], however, they could be any type of link:{spec_doc_url}#workflow-functions[functions], i.e REST, OpenAPI, gRPC, etc.
+For more details about error handling see xref:core/understanding-workflow-error-handling.adoc[Understanding workflow error handling].
+
+[#custom-function]
+.Example custom function using a Java class and method
+[source,json]
+----
+{
+   "name":"reserveStock",
+   "type":"custom",
+   "operation":"service:org.kie.kogito.PaymentService::processPayment"
+}
+----
+
+Kogito workflow engine controls all steps that were executed and should be compensated, in a stateful manner, this means the Saga can be long-lived, containing wait states among on the steps in the workflow, this means after each wait state the workflow is persisted and can continue in a different request, but for simplicity, this example shows the Saga as a request-response, also called straight-through workflow.
```

Comment:
I'd rewrite this paragraph. There are too many commas, and it's too long. It isn't easy to understand.

Reply:
+1